Add optional setting of pubsub credentials

pgalic96 opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes #1947 

#### Changes
<!-- What are the changes made in this pull request? -->

- Add conditional requirement based on the value of `secure` field in Pubsub integration
- Update generation of server url for nats

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

@bafonins pls check if the nats server url is generated correctly, and if mqtt pubsub integration is working correctly since I had some issues with creating one regardless of these changes

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

adriansmares commented:
I believe the above mentioned issue comes from the non-uniform ways the username and password are setup in the provider.
- The NATS provider is rather simple - it uses only the `ServerURL`, which follows basically the format of `nats://user:password@server:port`. I believe the swap between the username and the password occurs while parsing the badly formed `ServerURL`. My suggestion here would be to require both the username and password to be provided if the NATS connection is marked as secure.
- The MQTT provider is a bit more refined and has the username and password fields separately. Nonetheless, I would expect the same behavior - either both of the fields are provided, or none are provided.

kschiffer commented:
> My suggestion here would be to require both the username and password to be provided if the NATS connection is marked as secure.

> * The MQTT provider is a bit more refined and has the username and password fields separately. Nonetheless, I would expect the same behavior - either both of the fields are provided, or none are provided.

So let's do a checkbox then `Use credentials`, which is activated by default and disables the fields when deactivated.

Otherwise LGTM.

pgalic96 commented:
Updated the PR to include `Use Credentials` checkbox as suggested by @kschiffer

bafonins commented on pkg/webui/console/components/pubsub-form/index.js:
```diff
@@ -69,23 +74,33 @@ export default class PubsubForm extends Component {
+      ? mapNatsServerUrlToFormValue(initialPubsubValue.nats.server_url)
+      : { password: undefined, username: undefined, secure: false }
+    const initialNatsSecure = initialIsNats ? secure : false
+    const initialUseCredentialsNats = initialNatsSecure ? Boolean(password && username) : true
```
see my review comment

```suggestion
    const initialUseCredentialsNats = initialNatsSecure ? Boolean(password || username) : true
```

bafonins commented on pkg/webui/console/components/pubsub-form/index.js:
```diff
@@ -69,23 +74,33 @@ export default class PubsubForm extends Component {
     const initialIsMqtt = update && 'mqtt' in initialPubsubValue
     const initialMqttSecure = initialIsMqtt ? initialPubsubValue.mqtt.use_tls : false
+    const initialUseCredentialsMqtt = initialIsMqtt
+      ? Boolean(initialPubsubValue.mqtt.username && initialPubsubValue.mqtt.password)
```
see my review comment

```suggestion
      ? Boolean(initialPubsubValue.mqtt.username || initialPubsubValue.mqtt.password)
```

pgalic96 commented:
> However, since credentials are not enforced by the backend it can happen that only username or password is specified via cli/api/sdk and then the edit form is in inconsistent state (the `use_credentials` is not checked while the `password` or `username` field is provided). @adriansmares such behavior seems incorrect to me, anything we can do with this?
 So it should be possible in the console as well to specify only username or only password?

bafonins commented:
No, I dont think so. Just take into account that this can happen (with the current backend validation) and address https://github.com/TheThingsNetwork/lorawan-stack/pull/1985#discussion_r382631818, https://github.com/TheThingsNetwork/lorawan-stack/pull/1985#discussion_r382632048

cc @adriansmares

pgalic96 commented:
Alright. Regarding this:

> Also can we reuse the `use_credentials` checkbox between providers?

This would conflict with the idea of having `Use Credentials` checkbox checked by default. For example say we have a pubsub with nats provider which doesn't have `Use Credentials` checked. When changing the provider, all the settings are reset to default, so `Use Credentials` would be expected to be set to default value. However this would be disabled since its value is reused. Also, when changing back the existing values would be used except for this checkbox. Let me know what you think.

kschiffer commented on CHANGELOG.md:
```diff
@@ -25,6 +25,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Handling multiple consequent updates of MQTT pubsub/webhook integrations in the Console.
 - Displaying total device count in application overview section when using device search in the Console
 - FQDN used for Backend Interfaces interoperability requests.
+- Username and password not required in pubsub integration if tls is enabled
```
```suggestion
- Changing username and password to be not required in pubsub integration
```

kschiffer commented:
> This would conflict with the idea of having `Use Credentials` checkbox checked by default. For example say we have a pubsub with nats provider which doesn't have `Use Credentials` checked. When changing the provider, all the settings are reset to default, so `Use Credentials` would be expected to be set to default value. However this would be disabled since its value is reused. Also, when changing back the existing values would be used except for this checkbox. Let me know what you think.

You can use a ref to the formik element to get the `setValues` method of it:
```js
const formRef = React.useRef(null)
…
const { setValues, state } = formRef.current
// then set the value when the provider is changed:
setValues({
  ...state.values,
  _use_credentials: true, // you'd need some logic here to determine a consistent value
 })
```
Don't know though if it's worth the hassle.

pgalic96 commented:
According to @htdvisser comment here https://github.com/TheThingsNetwork/lorawan-stack/issues/2045#issuecomment-590719772 I think we need to allow setting only username for the mqtt. Not sure about nats? cc @bafonins 

Also I think reusing the `use_credentials` is not worth the hassle, but if you deem it important, I will work on fixing it.

pgalic96 commented:
Fixed the PR to include optional setting of username, also fixed an issue where unsetting `use_credentials` would still make an update request with the previously set username and password

bafonins commented on pkg/webui/console/components/pubsub-form/index.js:
```diff
@@ -69,23 +74,34 @@ export default class PubsubForm extends Component {
       error: '',
       isMqtt: initialIsMqtt,
       mqttSecure: initialMqttSecure,
+      natsSecure: initialNatsSecure,
```
What is the purpose of this entry? You never read it from the state

bafonins commented on pkg/webui/console/components/pubsub-form/index.js:
```diff
@@ -112,6 +128,14 @@ export default class PubsubForm extends Component {
   }
 
+  handleNatsSecureChange(event) {
+    this.setState({ natsSecure: event.target.checked })
```
What is the purpose of this entry? You never read it from the state

pgalic96 commented:
> 1. `mqtt.use_credentials` should not be sent to the backend

Fixed, not using this in the mapping function to determine what to send to server.
> 
> 1. Try this when editing mqtt pubsub:
> 
> * Check the `use_credentials` checkbox
> * Provide username and password, submit
> * Check the `use_credentials` checkbox again and submit
> * Check the `use_credentials` checkbox again, submit and see error
This (and the first issue) is fixed by setting values to null instead of stripping them, this issue was caused by casting the schema, but now I removed the schema casting, and the correct values are sent to backend and there is no need to cast the schema.
> 1. Why is only username required with mqtt provider?

Did you read my comment from yesterday? We also discussed it yesterday at the meeting



> According to @htdvisser comment here [#2045 (comment)](https://github.com/TheThingsNetwork/lorawan-stack/issues/2045#issuecomment-590719772) I think we need to allow setting only username for the mqtt. Not sure about nats? cc @bafonins
> 
> Also I think reusing the `use_credentials` is not worth the hassle, but if you deem it important, I will work on fixing it.

pgalic96 commented:
@bafonins rebased and squashed the fixup comments
